Kudu.Core/Extensions/JobManagerBaseIReadOnlyCollectionExtensions.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Kudu.Contracts.Jobs;
+using Kudu.Contracts.Settings;
+
+namespace Kudu.Core.Extensions
+{
+    public static class JobManagerBaseIReadOnlyCollectionExtensions
+    {
+        public static IJobsManager<T> GetPrimaryManager<T>(this IReadOnlyCollection<IJobsManager<T>> collection) where T : JobBase, new()
+        {
+            return collection.FirstOrDefault(m => m.BasePathType == WebJobsBasePathType.Primary);
+        }
+
+        public static IJobsManager<T> GetSecondaryManager<T>(this IReadOnlyCollection<IJobsManager<T>> collection) where T : JobBase, new()
+        {
+            return collection.FirstOrDefault(m => m.BasePathType == WebJobsBasePathType.Secondary);
+        }
+
+        public static IContinuousJobsManager GetWriteManager(this IReadOnlyCollection<IContinuousJobsManager> colleciton, IDeploymentSettingsManager settings)
+        {
+            return settings.RunFromLocalZip() || settings.RunFromRemoteZip()
+                ? colleciton.GetSecondaryManager() as IContinuousJobsManager
+                : colleciton.GetPrimaryManager() as IContinuousJobsManager;
+        }
+
+        public static ITriggeredJobsManager GetWriteManager(this IReadOnlyCollection<ITriggeredJobsManager> colleciton, IDeploymentSettingsManager settings)
+        {
+            return settings.RunFromLocalZip() || settings.RunFromRemoteZip()
+                ? colleciton.GetSecondaryManager() as ITriggeredJobsManager
+                : colleciton.GetPrimaryManager() as ITriggeredJobsManager;
+        }
```

Comment:
C# generics are not cooperating with me :) I'm not sure what the right syntax for this should be. The problem is that 

```cs
interface ITriggeredJobsManager : IJobsManager<TriggeredJob> { }
interface IContinuousJobsManager : IJobsManager<ContinuousJob> { }
```

Conceptually what I want is 

```cs
public static T GetWriteManager<T>(this IReadOnlyCollection<T> colleciton) where T : ITriggeredJobsManager, IContinuousJobsManager 
```

but because the interfaces inherit from an interface that takes a concrete type the definition won't compile because `m.BasePathType` is ambiguous between `IJobsManager<TriggeredJob>` and ` IJobsManager<ContinuousJob>`

Reply:
Feels like something that should be solvable, but I was not able to figure this out after playing with it for a bit.

Reply:
I know. I tried a bunch of stuff too and couldn't get it to work. I spent long enough on it then just gave up

```cs
public static T GetWriteManager<T, U>(this IReadOnlyCollection<IJobsManager<U>> c) 
    where U : JobBase, new()
```

```cs
public static T GetWriteManager<T, U>(this IReadOnlyCollection<IJobsManager<U>> c) 
    where U : TriggeredJob, ContinuousJob
```

```cs
public static T GetWriteManager<T>(this IReadOnlyCollection<T> c) 
    where T : ITriggeredJobsManager, IContinuousJobsManager 
```

Maybe we can create a wrapper type instead of an `IReadOnlyCollection`? something like

```cs
class WebJobManagerContainer
{
    public IContinuousJobsManager PrimaryContinuousManager { get; set; }
    public IContinuousJobsManager SecondaryContinuousManager { get; set; }

    public ITriggeredJobsManager PrimaryTriggeredManager { get; set; }
    public ITriggeredJobsManager SecondaryTriggeredManager { get; set; }
}
```

I guess that'll make it more explicit when compared to just a list.